fendermint/vm/topdown/src/finality/null.rs

```diff
@@ -149,6 +132,41 @@ impl FinalityWithNull {
 
 /// All the private functions
 impl FinalityWithNull {
+    /// Get the first non-null block in the range [start, end].
+    fn min_nonnull_block(&self, start: BlockHeight, end: BlockHeight) -> Stm<Option<BlockHeight>> {
+        let cache = self.cached_data.read()?;
+        for h in start..=end {
+            if let Some(Some(_)) = cache.get_value(h) {
+                return Ok(Some(h));
+            }
+        }
+        Ok(None)
+    }
+
+    fn propose_next_height(&self) -> Stm<Option<BlockHeight>> {
+        let latest_height = if let Some(h) = self.latest_height()? {
+            h
+        } else {
+            tracing::debug!("no proposal yet as height not available");
+            return Ok(None);
+        };
+
+        let last_committed_height = if let Some(h) = self.last_committed_finality.read_clone()? {
+            h.height
+        } else {
+            unreachable!("last committed finality will be available at this point");
+        };
+        let next_proposal_height = last_committed_height + self.min_proposal_interval;
+
+        Ok(Some(if next_proposal_height < latest_height {
+            // safe to unwrap as we are sure `latest_height` will not be null block
+            self.min_nonnull_block(next_proposal_height, latest_height)?
```

Comment:
Let's see if I get this right:
* `last_committed_height` is the last finalized block from the ledger
* `latest_height` is the highest record inserted into the cache; the fetcher guarantees that there is at least `chain_delay` blocks between that and the tip of the parent chain
* hopefully it is also not more than some limited distance away from `last_committed_height` as discussed in https://github.com/consensus-shipyard/fendermint/issues/437

Why should we impose on ourselves a `min_proposal_interval` when we can propose anything betweeen `last_committed_height` and `latest_height`? Going _back_ from `latest_height` to find a non-null block should be perfectly fine.

I suspect this exists because `latest_height` is still not limited in its look-ahead, and so you want to search something going from `last_committed_height` forward, but don't want to propose 1-by-1. 

Once again a `max_proposal_interval` would make more sense to me: it limits the potential blocks that have to be fetched and executed in one block, and when in sync it allows proposals to go 1-by-1, because the root chain is so much slower.

Reply:
I think `Going back from latest_height to find a non-null block should be perfectly fine.` might not work. Because other peers might be much slower, so the proposed height will always be rejected by peers. Starting from `last_committed_height` should be safer, even if someone is way ahead, peers should eventually catch up and the proposal will be accepted.

Reply:
> might not work. Because other peers might be much slower, so the proposed height will always be rejected by peers.

I agree, which is why I keep referring to https://github.com/consensus-shipyard/fendermint/issues/437
where I said that we should limit the `latest_height` by not fetching more than e.g. 100 ahead from `last_committed_height`, and why I have in this thread multiple times suggested replacing this new `min_topdown_interval` with a `max_topdown_interval`, which achieves almost the same thing, except in the linked issue I would also limit the maximum size of the cache that anyone would fetch.